### PR TITLE
Merge THStorage into at::Storage

### PR DESCRIPTION
**Pull request by cpuhrsch**

### CODE REVIEW

File: aten/src/ATen/Storage.cpp [hunk 1]
```diff
@@ -1,12 +1,17 @@
...
+
+Storage::Storage(
+    at::Backend backend,
     at::ScalarType scalar_type,
     ptrdiff_t size,
     at::DataPtr data_ptr,
     at::Allocator* allocator,
     char flag)
     : scalar_type(scalar_type),
+      backend_(backend),
```
Comment (ezyang) [inline aten/src/ATen/Storage.cpp]: Technically, you already know the backend from the contents of `data_ptr`, which has a `Device`.

File: aten/src/ATen/Storage.h [hunk 2]
```diff
@@ -1,53 +1,122 @@
...
+//
+//  - Once refcount == 0, it can never again be > 0 (the transition
+//    from > 0 to == 0 is monotonic)
+//
+//  - When you access THStorage via a weak pointer, you must
+//    atomically increment the use count, if it is greater than 0.
+//    If it is not, you must report that the storage is dead.
+//
+
+struct THFinalizer {
```
Comment (ezyang) [inline aten/src/ATen/Storage.h]: Now that it's not in TH anymore, give it a rename?

File: aten/src/ATen/Storage.h [hunk 3]
```diff
@@ -1,53 +1,122 @@
...
-  Storage() {}
-  Storage(THStorage* storage)
-      : storage(storage) {}
-  Storage(const Storage& other) = delete;
-  virtual ~Storage() {
-    THStorage_free(storage);
+  Storage() = delete;
+  virtual ~Storage() {};
+  Storage(at::Backend, at::ScalarType, ptrdiff_t, at::DataPtr, at::Allocator*, char);
+  Storage(at::Backend, at::ScalarType, ptrdiff_t, at::Allocator*, char);
```
Comment (ezyang) [inline aten/src/ATen/Storage.h]: Nit: can you add spaces here, so it's easier to visually distinguish between constructors, members and functions?

File: aten/src/ATen/Storage.h [hunk 4]
```diff
@@ -1,53 +1,122 @@
...
+  at::ScalarType scalar_type;
+  at::Backend backend_;
+  at::DataPtr data_ptr;
+  ptrdiff_t size;
+  std::atomic<int> refcount;
+  std::atomic<int> weakcount;
+  char flag;
+  at::Allocator* allocator;
+  std::unique_ptr<THFinalizer> finalizer;
+  struct Storage* view;
```
Comment (ezyang) [inline aten/src/ATen/Storage.h]: This field is dead, you can remove it.

File: aten/src/ATen/Storage.h [hunk 5]
```diff
@@ -1,53 +1,122 @@
...
+  template <typename T>
+  inline T* unsafe_data() const {
+    return static_cast<T*>(this->data_ptr.get());
+  }
+  static const char REFCOUNTED = 1;
+  static const char RESIZABLE = 2;
+
+//  virtual ~Storage() {
+//    THStorage_free(storage);
+//  }
```
Comment (ezyang) [inline aten/src/ATen/Storage.h]: Delete me?

File: aten/src/ATen/Storage.h [hunk 6]
```diff
@@ -1,53 +1,122 @@
...
+  Type& type() const;
+
   int getDevice() const {
-    return storage->data_ptr.device().index();
+    return data_ptr.device().index();
   }
-  void clear_flag(char flag) {
-    THStorage_clearFlag(storage, flag);
+  void clear_flag(char flag_) {
+    flag &= ~flag_;
```
Comment (ezyang) [inline aten/src/ATen/Storage.h]: Didn't you have a patch that converted this into a boolean?

File: aten/src/ATen/Utils.h [hunk 7]
```diff
@@ -24,9 +24,9 @@ AT_API int _crash_if_asan(int);
 
 template <typename T, typename Base>
 static inline T* checked_cast_storage(Base* expr, const char * name, int pos) {
-  if (typeid(*expr) != typeid(T))
-    AT_ERROR("Expected object of type ", T::typeString(), " but found type ", expr->type().toString(),
-             " for argument #", pos, " '", name, "'");
+//  if (typeid(*expr) != typeid(T))
+//    AT_ERROR("Expected object of type ", T::typeString(), " but found type ", expr->type().toString(),
+//             " for argument #", pos, " '", name, "'");
```
Comment (ezyang) [inline aten/src/ATen/Utils.h]: Uhhh... what's this? This can't possibly be right. I'm pretty sure you want something like https://github.com/pytorch/pytorch/pull/9764 here

Comment (cpuhrsch) [reply aten/src/ATen/Utils.h]: Yes, I did this to compile it and see what happens

File: aten/src/THC/THCGeneral.h.in [hunk 8]
```diff
@@ -64,6 +57,61 @@ typedef struct _THCCudaResourcesPerDevice {
   size_t scratchSpacePerStream;
 } THCCudaResourcesPerDevice;
 
+
+struct THCRNGState;  /* Random number generator state. */
+typedef struct CUDAStreamInternals THCStream;
+typedef struct THCState THCState;
+/* Global state of THC. */
```
Comment (ezyang) [inline aten/src/THC/THCGeneral.h.in]: Why did you move this? Likely to conflict with #9735.

File: aten/src/TH/generic/THStorage.h [hunk 9]
```diff
@@ -2,6 +2,8 @@
 #define TH_GENERIC_FILE "generic/THStorage.h"
 #else
 
+#include <ATen/Storage.h>
```
Comment (ezyang) [inline aten/src/TH/generic/THStorage.h]: This file is supposed to be C compatible, so this has to be protected with `ifdef __cplusplus`

File: aten/src/TH/generic/THStorage.h [hunk 10]
```diff
@@ -24,7 +26,9 @@
 #define TH_STORAGE_RESIZABLE  2
 
 // Struct definition is moved to THStorage.hpp (so this file stays C compatible)
-typedef struct THStorage THStorage;
+// typedef struct THStorage THStorage;
+
+#define THStorage at::Storage
```
Comment (ezyang) [inline aten/src/TH/generic/THStorage.h]: You need to vary this based on `__cplusplus` as well. Something like:

```
#ifdef __cplusplus
#define THStorage at::Storage
#else
typedef struct at_Storage at_Storage;
#define THStorage at_Storage
#endif
```

File: aten/src/ATen/templates/TypeDerived.cpp [hunk 11]
```diff
@@ -65,7 +65,7 @@ Tensor ${Type}::unsafeTensorFromTH(void * th_pointer, bool retain) const {
 std::unique_ptr<Storage> ${Type}::unsafeStorageFromTH(void * th_pointer, bool retain) const {
   if (retain)
     ${THStorage}_retain(${state,} (${THStorage}*) th_pointer);
-  return std::unique_ptr<Storage>(new ${Storage}((${THStorage}*) th_pointer));
+  return std::unique_ptr<Storage>((${THStorage}*) th_pointer);
```
Comment (ezyang) [inline aten/src/ATen/templates/TypeDerived.cpp]: This looks wrong. The standard deleter for a `unique_ptr` will call `delete` on the enclosed object. So when `std::unique_ptr<Storage>` becomes dead it will delete it. But what this function used to do is return a `unique_ptr` to Storage, **which is owning only for a single decrement of the reference count**; i.e., when the unique pointer dies, we merely decrement the refcount, and free the storage only if it's dead.

This interface is bad and confusing, but to fix it you would have to first refactor the `storage()` interface to not use unique pointers anymore, and have some sort of wrapper class `Storage` to `StorageImpl` (the same way we have `Tensor` to `TensorImpl`). I would prefer not to have to do it in this PR, but I am not sure if there is a way to fix this without changing the interface of unsafeStorageFromTH

How many use sites of unsafeStorageFromTH are there?

File: aten/src/ATen/templates/Tensor.h [hunk 12]
```diff
@@ -17,7 +17,7 @@
 
 namespace at {
 struct Type;
-struct Tensor;
+// struct Tensor;
```
Comment (ezyang) [inline aten/src/ATen/templates/Tensor.h]: Delete me

File: aten/src/ATen/templates/TensorDense.cpp [hunk 13]
```diff
@@ -6,7 +6,7 @@ Scalar ${Tensor}::localScalar() {
   return Scalar(${to_at_type}(${THStorage}_get(${state,} THTensor_getStoragePtr(tensor), tensor->storage_offset())));
 }
 std::unique_ptr<Storage> ${Tensor}::storage() {
-  auto storage = THTensor_getStoragePtr(tensor);
+  Storage* storage = THTensor_getStoragePtr(tensor);
   THStorage_retain(storage);
-  return std::unique_ptr<Storage>(new ${Storage}(storage));
+  return std::unique_ptr<Storage>(storage);
```
Comment (ezyang) [inline aten/src/ATen/templates/TensorDense.cpp]: This is also wrong, for the same reasons as in https://github.com/pytorch/pytorch/pull/9772/files#r204907148

**Review by ezyang — CHANGES_REQUESTED** [review]
There are some memory errors in this patch, fix those first and if your stated error persists we can debug more ;)

File: aten/src/THC/THCGeneral.h.in (hunk 8 above)
Comment (cpuhrsch) [reply aten/src/THC/THCGeneral.h.in]: Will move this back later if possible. Did this to deal with a forward declaration.

File: aten/src/ATen/Storage.h (hunk 2 above)
Comment (cpuhrsch) [reply aten/src/ATen/Storage.h]: Will do towards the end to keep things minimal.

File: aten/src/ATen/Storage.h (hunk 6 above)
Comment (cpuhrsch) [reply aten/src/ATen/Storage.h]: Might rebase on top of that or the other way around.

File: aten/src/ATen/templates/TypeDerived.cpp (hunk 11 above)
Comment (cpuhrsch) [reply aten/src/ATen/templates/TypeDerived.cpp]: There appears to be only two callsites

```
torch/csrc/DynamicTypes.cpp:153:  return type.unsafeStorageFromTH(((THPVoidStorage*)obj)->cdata, true);
torch/csrc/autograd/generated/VariableType.cpp:130:  return baseType->unsafeStorageFromTH(th_pointer, retain);
```

**Review by gchanan — COMMENTED** [review]
you have onnx changes, I assume you don't want that.

**Comment by cpuhrsch** [comment]
@gchanan - removed

File: aten/src/ATen/Storage.h [hunk 14]
```diff
@@ -1,53 +1,44 @@
 #pragma once
 
-#include "ATen/Scalar.h"
-#include <TH/THStorageFunctions.hpp>
+#include <ATen/StorageImpl.h>
 
 namespace at {
 
-struct Type;
+// OPtions: Use retainable
```
Comment (gchanan) [inline aten/src/ATen/Storage.h]: nit: OPtions

File: aten/src/ATen/Storage.h [hunk 15]
```diff
@@ -1,53 +1,44 @@
...
+// StorageImpl/StorageImplImpl thing
+// Inherit from retainable
+// StorageImpl has deleter, ; assignment and move deleted, pretty much like StorageImpl and THStorageImpl before
+// - - don't have to deleted unique ptr - will use intrusive and retainable
+
+
+// Details:
+// Forward the constructor just like before
+// Deconstructor just calls free on StorageImpl
+// Create a StorageImpl free function
```
Comment (gchanan) [inline aten/src/ATen/Storage.h]: I don't know what any of this means, but I assume you are going to delete it.

File: aten/src/ATen/Storage.h [hunk 16]
```diff
@@ -1,53 +1,44 @@
...
+    storage_impl_ = storage_impl;
+    if (storage_impl_)
+      storage_impl_->retain();
   };
+  ~Storage();
+  Storage(Storage&) = delete;
+  Storage(const Storage&) = delete;
+  Storage(Storage&&) = delete;
+  Storage(const Storage&&) = delete;
+  StorageImpl* storage_impl_;
```
Comment (gchanan) [inline aten/src/ATen/Storage.h]: why is this public?

File: aten/src/ATen/Storage.cpp [hunk 17]
```diff
@@ -0,0 +1,24 @@
...
+  }
+  if (storage_impl_->flag & StorageImpl::REFCOUNTED) {
+    if (--storage_impl_->refcount == 0) {
+      if (storage_impl_->finalizer) {
+        (*storage_impl_->finalizer)();
+      }
+      storage_impl_->finalizer = nullptr;
+      storage_impl_->data_ptr.clear();
+      if (storage_impl_ && --storage_impl_->weakcount == 0) {
+        storage_impl_ = nullptr;
```
Comment (gchanan) [inline aten/src/ATen/Storage.cpp]: how could setting a raw pointer to nullptr at the end of this destructor ever cause anything effect?

File: aten/src/ATen/StorageImpl.cpp [hunk 18]
```diff
@@ -15,14 +18,22 @@ THStorage::THStorage(
...
-    : THStorage(
+    : StorageImpl(
           scalar_type,
           size,
           allocator->allocate(at::elementSize(scalar_type) * size),
           allocator,
           flag) {}
+
+Type& StorageImpl::type() const {
+  if (data_ptr.device().is_cuda())
```
Comment (gchanan) [inline aten/src/ATen/StorageImpl.cpp]: Does this work with variables?  It looks like this is only used in one place anyway, it's probably just better to do exactly what you need there rather than implement this generally.

File: aten/src/ATen/StorageImpl.h [hunk 19]
```diff
@@ -0,0 +1,129 @@
...
+
+// Note [Weak references for intrusive refcounting]
+// ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+// Here's the scheme:
+//
+//  - refcount == number of strong references to the object
+//    weakcount == number of weak references to the object,
+//      plus one more if refcount > 0
+//
+//  - THStorageImpl stays live as long as there are any strong
```
Comment (gchanan) [inline aten/src/ATen/StorageImpl.h]: is THStorageImpl a thing?

File: aten/src/ATen/templates/TensorDense.cpp [hunk 20]
```diff
@@ -6,7 +6,7 @@ Scalar ${Tensor}::localScalar() {
   return Scalar(${to_at_type}(${THStorage}_get(${state,} THTensor_getStoragePtr(tensor), tensor->storage_offset())));
 }
 std::unique_ptr<Storage> ${Tensor}::storage() {
-  auto storage = THTensor_getStoragePtr(tensor);
+  StorageImpl* storage = THTensor_getStoragePtr(tensor);
```
Comment (gchanan) [inline aten/src/ATen/templates/TensorDense.cpp]: why?

File: aten/src/ATen/StorageImpl.h (hunk 19 above)
Comment (cpuhrsch) [reply aten/src/ATen/StorageImpl.h]: No, sorry, I'll delete all this garbage.

File: aten/src/ATen/templates/TensorDense.cpp (hunk 20 above)
Comment (cpuhrsch) [reply aten/src/ATen/templates/TensorDense.cpp]: I don't like auto.

File: aten/src/ATen/StorageImpl.cpp (hunk 18 above)
Comment (cpuhrsch) [reply aten/src/ATen/StorageImpl.cpp]: Sounds good.

File: aten/src/ATen/Storage.cpp (hunk 17 above)
Comment (cpuhrsch) [reply aten/src/ATen/Storage.cpp]: True

File: aten/src/ATen/Storage.h (hunk 16 above)
Comment (cpuhrsch) [reply aten/src/ATen/Storage.h]: I can add an accessor. It's easy to do in upcoming commits.

File: aten/src/ATen/templates/TensorDense.cpp (hunk 20 above)
Comment (gchanan) [reply aten/src/ATen/templates/TensorDense.cpp]: That's not a valid reason for changing the history.

File: aten/src/ATen/Storage.h [hunk 21]
```diff
@@ -1,53 +1,44 @@
...
 
-#include "ATen/Scalar.h"
-#include <TH/THStorageFunctions.hpp>
+#include <ATen/StorageImpl.h>
 
 namespace at {
 
-struct Type;
+// OPtions: Use retainable
+// Use strorage ptr == unique ptr <custom destructor calls release on storage>
```
Comment (ezyang) [inline aten/src/ATen/Storage.h]: strorage

File: aten/src/ATen/Storage.h [hunk 22]
```diff
@@ -1,53 +1,44 @@
...
+//  - custom destructor will just decrease refcount
+// Could rename this to StorageImplImpl and then create StorageImpl that maintains StorageImplImpl and decrease refcount
```
Comment (ezyang) [inline aten/src/ATen/Storage.h]: This comment should be deleted.

File: aten/src/ATen/Storage.h [hunk 23]
```diff
@@ -1,53 +1,44 @@
...
+
+struct Storage {
+  Storage() : storage_impl_(nullptr){};
+  Storage(StorageImpl* storage_impl) {
+    storage_impl_ = storage_impl;
+    if (storage_impl_)
+      storage_impl_->retain();
   };
+  ~Storage();
+  Storage(Storage&) = delete;
```
Comment (ezyang) [inline aten/src/ATen/Storage.h]: Worth a note that this is "temporary", in the sense that there are reasonable implementations of the copy/move constructors/assignment, we just didn't bother putting them in this PR.

File: aten/src/ATen/StorageImpl.cpp (hunk 18 above)
Comment (ezyang) [reply aten/src/ATen/StorageImpl.cpp]: Yeah, I definitely don't want this method on StorageImpl

File: aten/src/ATen/StorageImpl.h (hunk 19 above)
Comment (ezyang) [reply aten/src/ATen/StorageImpl.h]: Don't delete the entire note, it's still valid

File: aten/src/ATen/StorageImpl.h [hunk 24]
```diff
@@ -0,0 +1,129 @@
...
+  virtual void operator()() = 0;
+  virtual ~THFinalizer() {};
+};
+
+namespace at {
+
+struct Type;
+
+
+// OPtions: Use retainable
```
Comment (ezyang) [inline aten/src/ATen/StorageImpl.h]: More comments that need revising or deleting

File: aten/src/ATen/StorageImpl.h [hunk 25]
```diff
@@ -0,0 +1,129 @@
...
+          at::toString(scalar_type_T));
+    }
+    return unsafe_data<T>();
+  }
+
+  template <typename T>
+  inline T* unsafe_data() const {
+    return static_cast<T*>(this->data_ptr.get());
+  }
+  static const char REFCOUNTED = 1;
```
Comment (ezyang) [inline aten/src/ATen/StorageImpl.h]: We don't want REFCOUNTED at all. Storages are always refcounted, period. If you're going to refactor the constants, might as well get rid of them ;)

File: aten/src/ATen/Utils.h (hunk 7 above)
Comment (ezyang) [reply aten/src/ATen/Utils.h]: This is still a thing.

File: aten/src/ATen/templates/StorageDerived.cpp [hunk 26]
```diff
@@ -4,25 +4,46 @@
...
+  storage_impl_ = new StorageImpl(
+      ScalarType::${ScalarName}, 
+      0,
+#if ${isCUDA}
+      globalContext().getTHCState()->cudaDeviceAllocator,
+#else
+      getTHDefaultAllocator(),
+#endif
+      StorageImpl::RESIZABLE);
+  storage_impl_->clear_flag(StorageImpl::RESIZABLE);
```
Comment (ezyang) [inline aten/src/ATen/templates/StorageDerived.cpp]: This looks wrong. If we used the normal allocator to allocate memory for the storage, it should be resizable. Non-resizable storages are when we are given a pointer directly from some external source.

File: aten/src/ATen/templates/StorageDerived.cpp [hunk 27]
```diff
@@ -4,25 +4,46 @@
...
+  storage_impl_ = new StorageImpl(
+      ScalarType::${ScalarName}, 
+      size,
+#if ${isCUDA}
+      globalContext().getTHCState()->cudaDeviceAllocator,
+#else
+      getTHDefaultAllocator(),
+#endif
+      StorageImpl::RESIZABLE);
+  storage_impl_->clear_flag(StorageImpl::RESIZABLE);
```
Comment (ezyang) [inline aten/src/ATen/templates/StorageDerived.cpp]: Ditto here

File: aten/src/ATen/templates/StorageDerived.cpp [hunk 28]
```diff
@@ -4,25 +4,46 @@
...
-  : Storage(nullptr) {
-  storage = ${THStorage}_newWithAllocator(${state,} size, allocator);
-  ${THStorage}_clearFlag(${state,} storage, TH_STORAGE_RESIZABLE);
+${Storage}::${Storage}(size_t size, Allocator* allocator) {
+  storage_impl_ = new StorageImpl(
+      ScalarType::${ScalarName}, 
+      size,
+      allocator,
+      StorageImpl::RESIZABLE);
+  storage_impl_->clear_flag(StorageImpl::RESIZABLE);
```
Comment (ezyang) [inline aten/src/ATen/templates/StorageDerived.cpp]: And ditto here

File: aten/src/ATen/templates/StorageDerived.cpp [hunk 29]
```diff
@@ -36,28 +57,23 @@ static int getPointerDevice(void* ptr) {
...
-  return sizeof(${ScalarType});
-}
-
-Type& ${Storage}::type() const {
-  return globalContext().getType(Backend::${Backend},ScalarType::${ScalarName});
+       ),
+     /* allocator */ nullptr,
+     StorageImpl::RESIZABLE
+    );
+  storage_impl_->clear_flag(StorageImpl::RESIZABLE);
```
Comment (ezyang) [inline aten/src/ATen/templates/StorageDerived.cpp]: Here, clearing the resizable flag is correct, but then it is perplexing why you didn't just pass in 0 for the flag in the first case. Better yet, if you make resizable a boolean you can just say `/* resizable */ false`.

File: aten/src/TH/generic/THStorage.h [hunk 30]
```diff
@@ -24,7 +25,14 @@
 #define TH_STORAGE_RESIZABLE  2
 
 // Struct definition is moved to THStorage.hpp (so this file stays C compatible)
-typedef struct THStorage THStorage;
+// typedef struct THStorage THStorage;
```
Comment (ezyang) [inline aten/src/TH/generic/THStorage.h]: delete this

File: test/cpp/api/rnn.cpp [hunk 31]
```diff
@@ -110,15 +110,15 @@ TEST_CASE("rnn") {
     LSTM model(2, 2);
     for (auto& v : model->parameters()) {
       float size = v->numel();
-      auto p = static_cast<float*>(v->data().storage()->data());
+      auto p = static_cast<float*>(v->data().storage()->storage_impl_->data());
```
Comment (ezyang) [inline test/cpp/api/rnn.cpp]: Why not just put the actual methods on Storage? Or even better, implement `->` overloading so you don't have to enumerate all of them.

Comment (ezyang) [reply test/cpp/api/rnn.cpp]: Ditto all the occurrences below.

**Comment by ezyang** [comment]
Build failure is real, and if you implement `->` overloading it will fix all of the occurrences, instead of having to hunt them all out one-by-one.

File: torch/csrc/DynamicTypes.cpp [hunk 32]
```diff
@@ -134,7 +134,8 @@ PyObject* createPyObject(const at::Storage& storage)
   auto type = getPyTypeObject(storage);
   auto obj = THPObjectPtr(type->tp_alloc(type, 0));
   if (!obj) throw python_error();
-  ((THPVoidStorage*)obj.get())->cdata = (THVoidStorage *)storage.unsafeGetTH(true);
+  storage.pImpl()->retain();
+  ((THPVoidStorage*)obj.get())->cdata = (THVoidStorage *)storage.pImpl();
```
Comment (ezyang) [inline torch/csrc/DynamicTypes.cpp]: This is awkward. Why not have a retain argument in pImpl?

**Review by ezyang — APPROVED** [review]
Looks great!

**Comment by ezyang** [comment]
you got build failures:

```
22:16:08 /var/lib/jenkins/workspace/aten/src/ATen/native/cuda/Gesv.cu(107): error: class "at::Storage" has no member "data"
22:16:08 
22:16:08 /var/lib/jenkins/workspace/aten/src/ATen/native/cuda/Gesv.cu(108): error: class "at::Storage" has no member "data"
22:16:08 
22:16:08 /var/lib/jenkins/workspace/aten/src/ATen/native/cuda/Gesv.cu(109): error: class "at::Storage" has no member "data"
22:16:08 
22:16:08 /var/lib/jenkins/workspace/aten/src/ATen/native/cuda/Gesv.cu(110): error: class "at::Storage" has no member "data"
22:16:08           detected during instantiation of "void at::native::applyGesv<scalar_t>(at::Tensor &, at::Tensor &, std::vector<int64_t, std::allocator<int64_t>>) [with scalar_t=double]" 
22:16:08 (134): here
22:16:08 
```
